odl/tomo/operators/xray_trafo.py

```diff
@@ -0,0 +1,241 @@
+            if not np.allclose(discr_dom.grid.stride[1:],
+                               discr_dom.grid.stride[:-1]):
+                raise ValueError('ASTRA does not support different pixel/voxel'
+                                 ' sizes per axis (got {}).'
+                                 ''.format(discr_dom.grid.stride))
+
+        self._geometry = geometry
+
+        # Create discretized space (operator range). Use same data space
+        # type as the domain.
+        # TODO: maybe use a ProductSpace structure
+        ran_uspace = FunctionSpace(geometry.params)
+        # CHECKME: Is this the right weight?
+        ran_dspace = discr_dom.dspace_type(geometry.grid.ntotal,
+                                           weight=geometry.grid.cell_volume,
+                                           dtype=discr_dom.dspace.dtype)
+
+        ran_interp = kwargs.pop('range_interpolation', 'nearest')
+        discr_ran = DiscreteLp(ran_uspace, geometry.grid, ran_dspace,
+                               interp=ran_interp, order=geometry.grid.order)
+        super().__init__(discr_dom, discr_ran, linear=True)
+
+        if backend == 'astra':
+            if isinstance(discr_dom.dspace, CudaNtuples):
+                self._backend = 'astra_cuda'
+            else:
+                self._backend = 'astra_cpu'
+        else:
+            self._backend = backend
+
+        self._adjoint = DiscreteXrayTransformAdjoint(self)
+
+    @property
+    def backend(self):
+        """Computational backend for this operator."""
+        return self._backend
+
+    @property
+    def geometry(self):
+        """Geometry of this operator."""
+        return self._geometry
+
+    def _call(self, inp):
+        """Call the transform on an input, producing a new vector.
+
+        Parameters
+        ----------
+        inp : `DiscreteLpVector`
+           Element in the domain of the operator to be forward projected
+
+        Returns
+        -------
+        out : `DiscreteLpVector`
+            Returns an element in the projection space
+        """
+        back, impl = self.backend.split('_')
+        if back == 'astra':
+            if impl == 'cpu':
+                return astra_cpu_forward_projector_call(inp, self.geometry,
+                                                        self.range)
+            elif impl == 'cuda':
+                return astra_gpu_forward_projector_call(inp, self.geometry,
+                                                        self.range)
+            else:
+                raise ValueError('unknown implementation {}.'.format(impl))
+        else:  # Should never happen
+            raise RuntimeError('backend support information is inconsistent.')
+
+    @property
+    def adjoint(self):
+        """Return the adjoint operator."""
+        return self._adjoint
+
+
+class DiscreteXrayTransformAdjoint(Operator):
+
+    """The adjoint of the discrete X-ray transform."""
+
+    def __init__(self, forward):
```

Comment:
I think these should be operators in their own sense and not "simply an adjoint", you could see occations (FBP) where you do not want a forward projector at all.